internal/pubsub/pubsub_test.go

```diff
@@ -227,6 +228,46 @@ func TestCancelReceive(t *testing.T) {
 	}
 }
 
+type blockingDriverSub struct {
+	driver.Subscription
+	inReceiveBatch chan int
+}
+
+func (b blockingDriverSub) ReceiveBatch(ctx context.Context, maxMessages int) ([]*driver.Message, error) {
+	b.inReceiveBatch <- 0
+	<-ctx.Done()
+	return nil, ctx.Err()
+}
+
+func TestCancelTwoReceives(t *testing.T) {
+	// We want to create the following situation:
+	// 1. Goroutine 1 calls Receive, obtains the lock (Subscription.mu),
+	//    and calls driver.ReceiveBatch, which hangs forever.
+	// 2. Goroutine 2 calls Receive.
+	// 3. The context passed to the Goroutine 2 call is canceled.
+	// We expect Goroutine 2's Receive to exit immediately. That won't
+	// happen if Receive holds the lock during the call to ReceiveBatch.
+	inReceiveBatch := make(chan int, 1)
+	s := pubsub.NewSubscription(blockingDriverSub{inReceiveBatch: inReceiveBatch})
+	go func() {
+		s.Receive(context.Background())
+		panic("Receive should never return")
```

Comment:
t.Fatal ?

Reply:
t.Fatal in a goroutine doesn't work -- I think it panics instead -- but I guess that's fine.
Done.